python/pyspark/pandas/groupby.py

```diff
@@ -3534,7 +3534,12 @@ def _reduce_for_stat_function(
             for label in psdf._internal.column_labels:
                 psser = psdf._psser_for(label)
                 input_scol = psser._dtype_op.nan_to_null(psser).spark.column
-                output_scol = sfun(input_scol)
+                if sfun.__name__ == "sum" and isinstance(
+                    psdf._internal.spark_type_for(label), StringType
+                ):
+                    output_scol = F.concat_ws("", F.collect_list(input_scol))
```

Comment:
can we sort by natural order? we have `compute.ordered_head` config. We can sort it by natural order, and perform `collect_list`.

Reply:
I think maybe this is different case from `head`?

In head, we do `sdf = sdf.orderBy(NATURAL_ORDER_COLUMN_NAME)` and compute the `sdf.limit(n)`, so that we can keep the order because `DataFrame.limit` doesn't shuffle the data.

But in this case, it shuffles the data again when computing the `collect_list` even after sorting the DataFrame by natural order in advance, so I think the order would not be guaranteed.

Please let me know if I missed something??

Reply:
since string columns are computed together with numerical ones, I think we have to compute strings' sum in an aggregation way:

```
F.concat_ws("", F.array_sort(
  F.collect_list(F.struct(NATURAL_ORDER_COLUMN_NAME, input_scol))
)
```

For struct type, array_sort sort elements by first field then second field, IIRC

Reply:
Yeah, then maybe we should extract the only string column from nested arrays to pass as arguments to `concat_ws`?

```
F.concat_ws(
    "",
    F.array_sort(
        F.collect_list(F.struct(NATURAL_ORDER_COLUMN_NAME, input_scol))
    ).getField(input_scol_name),
)
```

Reply:
Just adjusted the comments. Thanks!

Reply:
yes, should extract the string col